src/tir/schedule/primitive/loop_transformation.cc

```diff
@@ -385,6 +511,116 @@ StmtSRef Fuse(ScheduleState self, const Array<StmtSRef>& loop_srefs) {
+  // - the top of the reorder range is the last loop visited in the first traversal which exists in
+  //   the input array
+  // - the bottom of the reorder range is the last loop in the input array which is not visited in
+  // the previous traversals
+  const StmtSRefNode* top = nullptr;
+  const StmtSRefNode* bottom = ordered_loop_srefs[0].get();
+  std::unordered_set<const StmtSRefNode*> visited;
+  bool scope_block_visited = false;
+  for (size_t i = 0; i < ordered_loop_srefs.size(); i++) {
+    const StmtSRefNode* loop_sref = ordered_loop_srefs[i].get();
+    if (visited.count(loop_sref)) {
+      continue;
+    }
+    for (const StmtSRefNode* v = loop_sref;; v = v->parent) {
+      // Case 1. If `v` corresponds to a block, stop traversal.
+      if (v->stmt->IsInstance<BlockNode>()) {
+        if (scope_block_visited) {
+          throw LoopsNotAChainError(self->mod, NullOpt,
+                                    LoopsNotAChainError::ProblemKind::kNotUnderAScope);
+        }
+        scope_block_visited = true;
+        break;
+      }
+      // Case 2. If `v` corresponds to a previously-visited loop, stop traversal and update
+      // `bottom`.
+      if (visited.count(v)) {
+        if (v == bottom) {
+          throw LoopsNotAChainError(self->mod, GetRef<Stmt>(v->stmt),
+                                    LoopsNotAChainError::ProblemKind::kHaveNonSingleBranchStmt);
+        }
+        bottom = loop_sref;
+        break;
+      }
+      // Case 3. Add `v` into `visited`
+      visited.insert(v);
+      // If it's the first traversal and the loop corresponding to `v` is in the input array,
+      // update `top`.
+      if (loop_srefs.count(v) && i == 0) {
+        top = v;
+      }
+    }
+  }
+  // Step 3. Collect all loops in the chain and check the loops are single-branch
+  std::vector<const StmtSRefNode*> chain;
+  for (const StmtSRefNode* loop_sref = bottom;; loop_sref = loop_sref->parent) {
+    if (!chain.empty()) {
+      const ForNode* outer_loop = TVM_SREF_TO_FOR(outer_loop, GetRef<StmtSRef>(loop_sref));
+      const ForNode* inner_loop = TVM_SREF_TO_FOR(inner_loop, GetRef<StmtSRef>(chain.back()));
+      if (outer_loop->body.get() != inner_loop) {
+        throw LoopsNotAChainError(self->mod, GetRef<For>(outer_loop),
+                                  LoopsNotAChainError::ProblemKind::kHaveNonSingleBranchStmt);
+      }
+    }
+    chain.push_back(loop_sref);
+    if (loop_sref == top) {
+      break;
+    }
+  }
+  // Step 4. Check the block below has all its block_var to be data-parallel or reduction,
+  // and the block has an affine binding.
+  BlockPropertyError::CheckBlockIterTypeAndAffineBinding(self, bottom);
+  // Step 5. Replace the original loops with the reordered loops and check that outer loop is
+  // not dependent on inner loop
+  std::unordered_set<const VarNode*> inner_vars;
+  For new_loop;
+  int index = ordered_loop_srefs.size() - 1;
```

Comment:
I know it is impossible for numerical underflow here, but we always treat `.size()` really carefully :-)

```suggestion
  int index = static_cast<int>(ordered_loop_srefs.size()) - 1;
```